### PR TITLE
More type inference for more binary expressions

### PR DESCRIPTION
**Pull request by marcusklaas**
Implements more of https://github.com/rust-analyzer/rust-analyzer/issues/390. Just works for primitive (numeric) types for now.

Found an issue where `let x: Ty = expr;` doesn't actually propagate the type information unless `Ty` is primitive and numeric. I'll open an issue for this.

### CODE REVIEW

File: crates/ra_hir/src/ty.rs [hunk 1]
```diff
@@ -529,15 +529,40 @@ struct InferenceContext<'a, D: HirDatabase> {
 
 // helper function that determines whether a binary operator
 // always returns a boolean
-fn is_boolean_operator(op: BinaryOp) -> bool {
+fn boolean_op_return_ty(op: BinaryOp, rhs_ty: Ty) -> Ty {
```
Comment (flodiebold) [inline crates/ra_hir/src/ty.rs]: this should be called `binary_op_return_ty`, right? it's not just for boolean ops. Also, the comment is outdated

File: crates/ra_hir/src/ty.rs [hunk 2]
```diff
@@ -889,20 +914,59 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
...
                             Expectation::has_type(Ty::Bool)
                         }
                         _ => Expectation::none(),
                     };
-                    let _lhs_ty = self.infer_expr(*lhs, &subtype_expectation)?;
-                    let _rhs_ty = self.infer_expr(*rhs, &subtype_expectation)?;
+                    let lhs_ty = self.infer_expr(*lhs, &lhs_expectation)?;
+                    // TODO: find implementation of trait corresponding to operation
+                    // symbol and resolve associated `Output` type
+                    let rhs_expectation = match op {
```
Comment (flodiebold) [inline crates/ra_hir/src/ty.rs]: maybe this big match could also be extracted into a function

File: crates/ra_hir/src/ty/tests.rs [hunk 3]
```diff
@@ -168,15 +168,18 @@ fn test() {
...
+    let g = minus_forty ^= i;
+    let ten: usize = 10;
+    let ten_is_eleven = ten == some_num;
 
-    10 < 3
+    ten < 3
 }
 "#,
-        "boolean_op.txt",
+        "binary_op.txt",
```
Comment (flodiebold) [inline crates/ra_hir/src/ty/tests.rs]: it seems you forgot to add the renamed file?

Comment (marcusklaas) [reply crates/ra_hir/src/ty/tests.rs]: uhh, I forgot to add it to git I think! :fearful:

File: crates/ra_hir/src/ty.rs (hunk 2 above)
Comment (marcusklaas) [reply crates/ra_hir/src/ty.rs]: Good idea. Will do.

File: crates/ra_hir/src/ty.rs (hunk 1 above)
Comment (marcusklaas) [reply crates/ra_hir/src/ty.rs]: Yes!

**Comment by marcusklaas** [comment]
Thanks again for the review and apologies for the sloppy mistakes - I will try to do better!

**Comment by flodiebold** [comment]
No problem, thank you :smile: 

bors r+

**Pull request merged**